lib/ansible/executor/process/threading.py

```diff
@@ -0,0 +1,170 @@
+# (c) 2018, Red Hat, Inc. <support@ansible.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+# Make coding more python3-ish
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os
+import sys
+import threading
+import time
+import traceback
+
+from collections import deque
+from jinja2.exceptions import TemplateNotFound
+
+from ansible.errors import AnsibleConnectionFailure
+from ansible.executor.process.model import ProcessModelBase, ResultsSentinel
+from ansible.executor.task_executor import TaskExecutor
+from ansible.executor.task_result import TaskResult
+from ansible.module_utils._text import to_text
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+__all__ = ['ProcessModelThreading', 'WorkerProcess']
+
+
+class ProcessModelThreading(ProcessModelBase):
+    def __init__(self, tqm):
+        super(ProcessModelThreading, self).__init__(tqm)
+
+        self._job_queue = deque()
+        self._job_queue_lock = threading.Lock()
+
+    def initialize_workers(self, num):
+        self._cur_worker = 0
+        self._workers = []
+        for i in range(num):
+            w_thread = threading.Thread(target=run_worker, args=(self, self._shared_loader_obj))
+            w_thread.start()
+            w_lock = threading.Lock()
+            self._workers.append([w_thread, w_lock])
+
+    def get_job(self):
+        data = None
+        try:
+            self._job_queue_lock.acquire()
+            data = self._job_queue.pop()
+        except:
+            pass
+        finally:
+            self._job_queue_lock.release()
+        return data
+
+    def put_task(self, data):
```

Comment:
Just as a note here, we call this `put_task` but getting the data from the queue we call it `get_job` which seems inconsistently named.

Reply:
Also, instead of using `appendleft` here and `pop` above, we could `append`, and `popleft`, it doesn't really matter.  Although I think we should order all of our queues so they basically operate the same.